src/pydvl/valuation/methods/gt_shapley.py

```diff
@@ -12,325 +12,375 @@
-            estimation of `n_iterations`.
-        delta: From the (ε,δ) sample bound. Use the same as for the
-            estimation of `n_iterations`.
-        n_jobs: Number of parallel jobs to use. Each worker performs a chunk
-            of all tests (i.e. utility evaluations).
-        parallel_backend: Parallel backend instance to use
-            for parallelizing computations. If `None`,
-            use [JoblibParallelBackend][pydvl.parallel.backends.JoblibParallelBackend] backend.
-            See the [Parallel Backends][pydvl.parallel.backends] package
-            for available options.
-        config: (**DEPRECATED**) Object configuring parallel computation,
-            with cluster address, number of cpus, etc.
-        progress: Whether to display progress bars for each job.
-        seed: Either an instance of a numpy random number generator or a seed for it.
-        options: Additional options to pass to
-            [cvxpy.Problem.solve()](https://www.cvxpy.org/tutorial/advanced/index.html#solve-method-options).
-            E.g. to change the solver (which defaults to `cvxpy.SCS`) pass
-            `solver=cvxpy.CVXOPT`.
+    problem = GroupTestingProblem(
+        utility_differences=u_differences,
+        total_utility=total_utility,
+        epsilon=epsilon,
+    )
 
-    Returns:
-        Object with the data values.
+    return problem
 
-    !!! tip "New in version 0.4.0"
 
-    !!! tip "Changed in version 0.5.0"
-        Changed the solver to cvxpy instead of scipy's linprog. Added the ability
-        to pass arbitrary options to it.
+def _calculate_utility_differences(
+    utility_values: NDArray[np.float_],
+    masks: NDArray[np.bool_],
+    n_obs: int,
+) -> NDArray[np.float_]:
+    """Calculate utility differences from utility values and sample masks.
+
+    Args:
+        utility_values: 1D array with utility values.
+        masks: 2d array with sample masks.
+        n_obs: The number of observations.
+
+    Returns:
+        The utility differences.
 
-    !!! tip "Changed in version 0.9.0"
-        Deprecated `config` argument and added a `parallel_backend`
-        argument to allow users to pass the Parallel Backend instance
-        directly.
     """
+    betas = masks.astype(np.int_)
+    n_samples = len(utility_values)
 
-    n = len(u.data.indices)
+    z = _calculate_z(n_obs)
 
-    const = _constants(
-        n=n,
-        epsilon=epsilon,
-        delta=delta,
-        utility_range=u.score_range.max() - u.score_range.min(),
-    )
-    T = n_samples
-    if T < const.T:
-        log.warning(
-            f"n_samples of {T} are below the required {const.T} for the "
-            f"ε={epsilon:.02f} guarantee at δ={1 - delta:.02f} probability"
-        )
+    u_differences = np.zeros(shape=(n_obs, n_obs))
+    for i in range(n_obs):
+        for j in range(i + 1, n_obs):
+            u_differences[i, j] = np.dot(utility_values, betas[:, i] - betas[:, j])
```

Comment:
I guess, this is not performance critical. Only if this is critical, one could think of vectorizing the double for loop, but maybe having all differences of beta in memory is not possible?

Reply:
Yes, basically only Utility evaluations are performance critical.